### PR TITLE
feat(lobbies): guard lobbies api with captcha

### PR DESCRIPTION
**Pull request by ABCxFF**

### CODE REVIEW

**Comment by ABCxFF** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/578?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#578** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/578?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#577** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/577?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#576** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/576?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ABCxFF and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: modules/lobbies/scripts/list.ts [hunk 1]
```diff
@@ -29,8 +32,16 @@ export async function run(
 	ctx: ScriptContext,
 	req: Request,
 ): Promise<Response> {
-	// TODO: Cache this without hitting the DO
+	await ctx.modules.captcha.guard({
+		key: "lobbies.list",
+		period: 20,
+		requests: 5,
+		type: "default",
```
Comment (NathanFlurry) [inline modules/lobbies/scripts/list.ts:39]: type & key are flipped here iirc if we want to keep this consistent with rate limit. can chat about a better api for this over slack.

File: modules/lobbies/scripts/list.ts [hunk 2]
```diff
@@ -29,8 +32,16 @@ export async function run(
 	ctx: ScriptContext,
 	req: Request,
 ): Promise<Response> {
-	// TODO: Cache this without hitting the DO
+	await ctx.modules.captcha.guard({
+		key: "lobbies.list",
+		period: 20,
```
Comment (NathanFlurry) [inline modules/lobbies/scripts/list.ts:37]: period & requests should be configurable in config.ts and have defaults in the default config

File: modules/lobbies/scripts/join.ts [hunk 3]
```diff
@@ -25,6 +28,15 @@ export async function run(
 	ctx: ScriptContext,
 	req: Request,
 ): Promise<Response> {
+	await ctx.modules.captcha.guard({
+		key: "lobbies.join",
+		period: 15,
```
Comment (NathanFlurry) [inline modules/lobbies/scripts/join.ts:33]: period & requests should be configurable in config.ts and have defaults in the default config

File: modules/lobbies/scripts/find_or_create.ts [hunk 4]
```diff
@@ -34,6 +37,15 @@ export async function run(
 	ctx: ScriptContext,
 	req: Request,
 ): Promise<Response> {
+	await ctx.modules.captcha.guard({
+		key: "lobbies.find_or_create",
+		period: 15,
```
Comment (NathanFlurry) [inline modules/lobbies/scripts/find_or_create.ts:42]: period & requests should be configurable in config.ts and have defaults in the default config

File: modules/lobbies/utils/captcha_config.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { Config, CaptchaProvider } from "../config.ts";
+
+export const getCaptchaProvider = (config: Config): CaptchaProvider => {
```
Comment (NathanFlurry) [inline modules/lobbies/utils/captcha_config.ts:3]: use functions not export const

File: modules/lobbies/scripts/find.ts [hunk 6]
```diff
@@ -27,6 +29,16 @@ export async function run(
 	ctx: ScriptContext,
 	req: Request,
 ): Promise<Response> {
+	await ctx.modules.captcha.guard({
+		key: "lobbies.find",
+		period: 15,
```
Comment (NathanFlurry) [inline modules/lobbies/scripts/find.ts:34]: period & requests should be configurable in config.ts and have defaults in the default config

File: modules/lobbies/scripts/create.ts [hunk 7]
```diff
@@ -33,6 +36,15 @@ export async function run(
 	ctx: ScriptContext,
 	req: Request,
 ): Promise<Response> {
+	await ctx.modules.captcha.guard({
+		key: "lobbies.create",
+		period: 15,
```
Comment (NathanFlurry) [inline modules/lobbies/scripts/create.ts:41]: period & requests should be configurable in config.ts and have defaults in the default config

File: modules/lobbies/utils/captcha_config.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+    if (!config.captcha || !config.captcha.provider) return {
```
Comment (NathanFlurry) [inline modules/lobbies/utils/captcha_config.ts:4]: a null captcha provider currently returns a test provider. we should not enforce captchas if there is no provider provided.

**Pull request closed without merging**